### ISSUE TITLE
Scrapy under Python 3 is slower than under Python 2

### ISSUE THREAD

**Issue by lopuhin** [scrapy/scrapy#3134]
bookworm benchmark from https://github.com/scrapy/scrapy-bench/ (see also https://medium.com/@vermaparth/parth-gsoc-f5556ffa4025) shows about 15% slowdown, while more synthetic ``scrapy bench`` shows a 2x slowdown: https://github.com/scrapy/scrapy/pull/3050#issuecomment-353863711

**Comment by cathalgarvey** [comment]
I think some Py3 performance issues are the penalty paid for sane Unicode handling, is that what's hitting Scrapy here? I wonder how much string-handling is happening in the plumbing, that could be handled as bytes instead for performance?

**Comment by lopuhin** [comment]
Yes, it's likely one of the reasons. I think first step would be to compare profiles under Py2 and Py3 for several benchmarks

**Comment by nctl144** [comment]
@lopuhin would you mind showing me which benchmarks are you mentioning?

**Comment by lopuhin** [comment]
@nctl144 I meant the ones present in this comment: https://github.com/scrapy/scrapy/issues/3134#issue-298868638 - basically this means "all of them", sorry. Probably the biggest difference I've personally seen is in ``scrapy bench``, but I didn't study this.

**Comment by lopuhin** [comment]
I think I would start from comparing profiles (that means running under profiler) under Py2.7 and Py3.6 for several benchmarks and trying to spot where most of the difference comes from. For benchmarks, I think it makes sense to check several of them, because some might show more difference and some will be easier to analyze. For profilers, if you already have some preference then go with it. If not, I would suggest using built-in cProfile with some visualization backend (e.g. snakeviz), and vmprof + vmprof.com for visualization - it's good to have several different profilers because this allows to cross-check profiling results.

**Comment by nctl144** [comment]
Hey @lopuhin, I made some profiling on Scrapy in python2 and python3 using scrapy-bench. The result is noted [here](https://docs.google.com/document/d/1TTuvZVf-AIxypwLwMQgBtFN6b7wkLRV9eJZ4TRT0R-w/edit?usp=sharing), and I will continue updating it after I've done more research on these profiles!

**Comment by lopuhin** [comment]
Interesting results @nctl144 ! Btw, this issue https://github.com/scrapy/scrapy/issues/1306 could be also relevant here.

**Comment by nctl144** [comment]
I've read through the discussion and it might be a good idea to record the mentioned options into consideration. I will continue to analyze Scrapy performance differences between 2 environments and update the document. I will let you know if there is any progress made on that.
Is there any suggestion that you can give me to move further with this topic?

**Comment by lopuhin** [comment]
> Is there any other suggestion that you can give me here to further develop the project idea?

Not really, I think first step is to measure and analyze the differences and where they come from, like hypothetically "urljoin (taking 20%) time overall is taking 100% more time under python 3, likely because of this and that". Also I think it makes sense to profile ``scrapy bench`` too - I remember the difference there was also quite large.

**Comment by nctl144** [comment]
@lopuhin Aight I got that! What I am doing now is concentrating on profiling the performance of `bookworm` and `scrapy bench ` on python2 and python3.
Profiling with Snakeviz is fine since they have a doc which describes the profiling result in detail. However, since this is the first time I use **vmprof** so I don't really know what information in the result page represents. I tried to look for the answers but I found none. Therefore, can you help me briefly explain the main parts that I need to pay attention to when looking at the vmprof result?
Thank you so much! 😄

**Comment by lopuhin** [comment]
@nctl144 sure, the main difference of vmprof is that it is a statistical profiler: https://en.wikipedia.org/wiki/Profiling_(computer_programming)#Statistical_profilers so it does not tell how many times a function was called, so it's useful to check cProfile results as well. But as you many have noticed, it has lower overhead, especially when there are a lot of calls of small and fast functions. 

Default mode of vmprof display is called a "flamegraph", but there are two more modes reachable by the buttons on the bottom left in the screenshot
<img width="601" alt="screen shot 2018-03-02 at 14 38 29" src="https://user-images.githubusercontent.com/424613/36897422-890d139a-1e27-11e8-9126-4ea67b611a5a.png">

There you can get %-age of time spent in the function (both "safe" and "cumulative").

**Comment by nctl144** [comment]
Hey @lopuhin , thank you so much for your help. I took a look at the graphs of bookworm scrapy-bench and I have some result. I just need your feedback on it just to know that I am going into the right direction.

|| Python2 | Python3
| ---- | --- |  ----
Parse  | 46.5% | 48.0%
_extract_request | 16.5% | 20.1%
extract_links | 13.3% | 16.9%
| genexp | 13.1% | 14.6%

So I suppose for those functions (possibly including **urlsplit, tokenize,...**, according to the detailed result that I recorded [here](https://docs.google.com/document/d/1TTuvZVf-AIxypwLwMQgBtFN6b7wkLRV9eJZ4TRT0R-w/edit?usp=sharing)) , python3 is considered to be slower, am I correct?
I drew my conclusion based on the result that xpath and cssbench in python2 are slower than in python3, and applied that to the graph of bookworm.

**Comment by lopuhin** [comment]
@nctl144 yes, I think your analysis is correct.

**Comment by nctl144** [comment]
Great, thank you for your feedback @lopuhin . I made another analysis on scrapy bench and I saw some interesting result with big differences between scrapy on python2 and on python3.
On python2, scrapy bench was able to crawl 653 pages (at 4320 pages/min)
On python3, scrapy bench was able to crawl 421 pages (at 2040 pages/min)
So it's about 30% slower on python3.
And here's some analysis on % cpu that I notice:

| | Python2 | Python3
-- | -- | --
Parse | 30.3% | 33.9%
extract_links | 18.8% | 29.6%
canonicalize_url | 15.6% | 27.1%
crawl | 13.7% | 15.5%
schedule | 13.5% | 15.4%

I think a part of this matches with the discussion in #1306 . What do you think?

**Comment by lopuhin** [comment]
> On python2, scrapy bench was able to crawl 653 pages (at 4320 pages/min)
> On python3, scrapy bench was able to crawl 421 pages (at 2040 pages/min)
> So it's about 30% slower on python3.

Could there be a typo somewhere, maybe it's 3040 rpm for python 3, or difference is not 30%?

But anyway, it seems that url parsing and handling is the major source of the slowdown compared to python 2, and it's worth fixing.

Let me think about it more and gather feedback from other scrapy developers.

**Comment by nctl144** [comment]
I dont think it's a typo, @lopuhin . I tried running **scrapy bench** many times on my computer and it gave the same result.
But please let me know what you are thinking of, so that I might be able to implement the potential fixes whose performance can be further analyzed.  I am interested!

**Comment by nctl144** [comment]
I suppose I will continue this discussion #1306 and try all the methods that are suggested in there first!

**Comment by nctl144** [comment]
Hey @lopuhin , I have tested some other urlparse library and this is the result I got:


  | Python2 | Python3
-- | -- | --
python urlparse | 0.113811016083 seconds | 0.17932915687561035 seconds 
purl | 0.549368143082 seconds | 0.5122480392456055 seconds
furl | 6.29960584641 seconds | 4.550176620483398 seconds
yurl | 0.105498075485 seconds | 0.10087418556213379 seconds

And this is my test code:
```
import time


start_time = time.time()

with open('urls.txt', 'r') as f:
    for url in f:
        a = URL(url)

print("--- %s seconds ---" % (time.time() - start_time))
```

Where **urls.txt** contains about 22,000 urls. 
I think yurl is pretty good since it lower the gap between the performance of scrapy on py2 and py3. I haven't tried other suggested libraries yet. But I just want to hear your opinion before I continue doing some more research about this.
Please let me know!

**Comment by lopuhin** [comment]
> I dont think it's a typo, @lopuhin . I tried running scrapy bench many times on my computer and it gave the same result.

I meant that the difference in pages/min is 2x (4320 vs 2040), while in total pagese crawls is about 30% (653  vs 421).

> I think yurl is pretty good since it lower the gap between the performance of scrapy on py2 and py3.

Yes, indeed your tests show that it's faster, this is great! Drawbacks that I currently see is that (1) there is not license I could find (https://github.com/homm/yurl/issues/2) and (2) it does not seem maintained. Some general thoughts:
- it could be valuable to check some other common operations, at least urljoin
- for libraries that are not pure python it's important to consider ease of installation, we don't want to make it more complicated (but a library that reliably provides wheels on common platforms is fine IMO).
- #1306 is also about making url parsing more correct - would be good to check how candidate libraries fare here

I think it would be better to continue discussion related to faster URL parsing in #1306.

**Comment by nctl144** [comment]
@lopuhin sounds great! I will check on what you suggested and continue discussing in #1306 !

**Comment by nctl144** [comment]
> I meant that the difference in pages/min is 2x (4320 vs 2040), while in total pagese crawls is about 30% (653 vs 421).

And you are right. Probably I was running something else that could affect scrapy bench while it was running (Or I just copied the wrong line). I have run it again and it gave the following result:
Python2:
`2018-03-13 01:30:58 [scrapy.extensions.logstats] INFO: Crawled 693 pages (at 2880 pages/min)`
Python3:
`2018-03-13 01:30:33 [scrapy.extensions.logstats] INFO: Crawled 452 pages (at 2400 pages/min)`

So this gave the expected result, which is: Scrapy on Python3 is about 20% slower on Python 2!
Thank you for pointing that out to me :)

**Comment by AdrianAtZyte** [comment]
Closing given Python 2 is dead.

But, for the record, we now have CodSpeed set up to catch performance regressions being introduced in PRs and to implement measurable performance improvements.
